server/core/src/main/java/dev/slimevr/osc/VRCOSCHandler.kt

```diff
@@ -0,0 +1,485 @@
+						val (w, x, y, z) = EulerAngles(EulerOrder.YXZ, event.message.arguments[0] as Float * FastMath.DEG_TO_RAD, event.message.arguments[1] as Float * FastMath.DEG_TO_RAD, event.message.arguments[2] as Float * FastMath.DEG_TO_RAD).toQuaternion()
+						receivingRotationOffsetGoal = Quaternion(w, -x, -y, z).inv()
+
+						receivingRotationOffsetGoal = if (slimeHead != null && slimeHead.hasRotation) {
+							slimeHead.getRotation().project(Vector3.POS_Y).unit() * receivingRotationOffsetGoal
+						} else {
+							receivingRotationOffsetGoal
+						}
+
+						// If greater than 300ms, snap to rotation
+						if (System.currentTimeMillis() - timeAtLastReceivedRotationOffset > 300) {
+							receivingRotationOffset = receivingRotationOffsetGoal
+						}
+
+						// Update time variable
+						timeAtLastReceivedRotationOffset = System.currentTimeMillis()
+					}
+				} else {
+					// Trackers data (1-8)
+					val trackerId = trackerStringValue[0].digitToInt()
+					var tracker = oscTrackersDevice!!.trackers[trackerId]
+
+					if (tracker == null) {
+						tracker = Tracker(
+							device = oscTrackersDevice,
+							id = VRServer.getNextLocalTrackerId(),
+							name = "OSC Tracker #$trackerId",
+							displayName = "OSC Tracker #$trackerId",
+							trackerNum = trackerId,
+							trackerPosition = null,
+							hasRotation = true,
+							hasPosition = true,
+							userEditable = true,
+							isComputed = true,
+							needsReset = true,
+							usesTimeout = true
+						)
+						oscTrackersDevice!!.trackers[trackerId] = tracker
+						server.registerTracker(tracker)
+					}
+
+					// Sets the tracker status to OK
+					tracker.status = TrackerStatus.OK
+
+					if (event.message.address.toString() == "/tracking/trackers/$trackerId/position") {
+						tracker.position = receivingRotationOffset.sandwich(
+							Vector3(
+								event.message.arguments[0] as Float,
+								event.message.arguments[1] as Float,
+								-(event.message.arguments[2] as Float)
+							) - receivingPositionOffset
+						) + postReceivingPositionOffset
+					} else {
+						val (w, x, y, z) = EulerAngles(EulerOrder.YXZ, event.message.arguments[0] as Float * FastMath.DEG_TO_RAD, event.message.arguments[1] as Float * FastMath.DEG_TO_RAD, event.message.arguments[2] as Float * FastMath.DEG_TO_RAD).toQuaternion()
+						val rot = Quaternion(w, -x, -y, z)
+						tracker.setRotation(receivingRotationOffset * rot * postReceivingOffset)
+					}
+
+					tracker.dataTick()
+				}
+			}
+		}
+	}
+
+	override fun update() {
+		// Update current time
+		val currentTime = System.currentTimeMillis().toFloat()
+
+		// Gets timer from vrServer
+		if (fpsTimer == null) {
+			fpsTimer = VRServer.instance.fpsTimer
+		}
+		// Update received trackers' offset rotation slerp
+		if (receivingRotationOffset != receivingRotationOffsetGoal) {
+			receivingRotationOffset = receivingRotationOffset.interpR(receivingRotationOffsetGoal, OFFSET_SLERP_FACTOR * (fpsTimer?.timePerFrame ?: 1f))
```

Comment:
Ideally, we shouldn't be using a constant interval for our slerp time, this will cause inconsistent movement speed. It really depends on the rate here whether that wouldn't matter or what, but I think we should instead use the time between the last received rotation and the current one, so it would take an equivalent time to interpolate as the time between the messages. If it's always consistent, it should stay smooth, but that's really only an ideal... The *"best"* solution would be introducing a buffer (latency) to prevent jitter, but that may not be ideal for our low-latency goal... I'd like some input on this.

Reply:
Using this kind of interpolation is intentional, I talked about it with Kung as I initially used my Quaternion filtering class (that does consistent smoothing/no acceleration) 
The design goal is that the rotation should not change if the user is moving their head a bit, but change more if they happen to do a 180. This is only used if constant head rotation is received, for debug or constant alignment (in which case the speed shouldn’t matter)

Reply:
Hmm okay. I don't quite get your explanation, but if there's solid reasoning behind this decision then that's fine.